botocore/waiter.py

```diff
@@ -309,23 +324,35 @@ def wait(self, **kwargs):
                     # can just handle here by raising an exception.
                     raise WaiterError(
                         name=self.name,
-                        reason=response['Error'].get('Message', 'Unknown'),
-                        last_response=response
+                        reason='An error occurred (%s): %s' % (
+                            response['Error'].get('Code', 'Unknown'),
+                            response['Error'].get('Message', 'Unknown'),
+                        ),
+                        last_response=response,
                     )
             if current_state == 'success':
                 logger.debug("Waiting complete, waiter matched the "
                              "success state.")
                 return
             if current_state == 'failure':
+                reason = 'Waiter encountered a terminal failure state: %s' % (
+                        acceptor.explanation
```

Comment:
This will potentially be None if we implement a new acceptor later and blow up with a `ValueError` on formatting. Should be set up a safer default in the event of no match?

Reply:
@nateprewitt In practice it's probably unlikely. If an unknown waiter type is encountered you can't actually instantiate the `AcceptorConfig` object anyways. Though I suppose a default message long the lines of `'No explanation for unknown waiter type "%s"'`.